### ISSUE TITLE
Vector Quantization Layer

### ISSUE THREAD

**Issue by Simsso** [Simsso/NIPS-2018-Adversarial-Vision-Challenge#51]
Development of a production-ready vector quantization (VQ) layer in TensorFlow, based on the prototype developed in #25 and merged with #52 (+prototype 2).

Development branch: [`vq-layer`](https://github.com/Simsso/NIPS-2018-Adversarial-Vision-Challenge/tree/vq-layer)

[**Documentation**](https://github.com/Simsso/NIPS-2018-Adversarial-Vision-Challenge/tree/vq-layer/vq-layer#vector-quantization-layer)

### Sub-tasks
* [x] Test batch size effect on gradient (https://github.com/Simsso/NIPS-2018-Adversarial-Vision-Challenge/commit/f39059513466282879e42f4a32243c7df6113557)
* [x] Assignment of least used embedding space vectors with values from input `x` that were furthest away from embedding space vectors
* [x] Consider developing a custom C++ op
* [x] Research on current memory efficiency (which parts consume a lot of RAM, how to profile)
* [x] `is_training` parameter updates only during training (not needed anymore)
* [x] Add `scatter_update` call to `tf.GraphKeys.UPDATE_OPS`
* [ ] `dotp` norm order #63
